### ISSUE TITLE
SA1312 : Variable '_' should begin with lower-case letter

### ISSUE THREAD

**Issue by cstalder** [DotNetAnalyzers/StyleCopAnalyzers#3057]
It should be an exception for '_'

**Comment by sharwell** [comment]
Can you give an example of where this would apply?

**Comment by cstalder** [comment]
Task<ICollection<ValidationError>> _ = ValidateAsync();

**Comment by sharwell** [comment]
You can change the above line to this to address the warning:

```csharp
_ = ValidateAsync();
```

**Comment by jltjohanlindqvist** [comment]
Hijacking this issue a bit, the solution you suggested @sharwell doesn't work for this case though:

```
using var _ = File.Open("asd", FileMode.Open);
```

**Comment by sharwell** [comment]
@jltjohanlindqvist I believe that case will eventually be addressed by dotnet/csharplang#2235. In the meantime, a named variable should be used.

**Comment by jltjohanlindqvist** [comment]
Alright, thanks for the quick answer!

**Comment by cnshenj** [comment]
@sharwell what about `foreach`?
```csharp
bool HasAtLeast(IEnumerable e, int i)
{
    var count = 0;
    foreach (var _ in e)  // SA1312 false positive
    {
        if (++count == i)
        {
            return true;
        }
    }

    return false;
}
```

**Comment by JoshuaLight** [comment]
Have the same problem with `foreach`. Is there any way to workaround?

**Comment by sharwell** [comment]
The `foreach` case is the same as https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3057#issuecomment-552514879

**Comment by cliedeman** [comment]
What about variables that start with a number?

```
_3MonthsFromNow
```
